### ISSUE TITLE
(ec2): Cannot attach Volume to Instance

### ISSUE THREAD

**Issue by drmoore718** [aws/aws-cdk#16001]
ec2.Instance only currently supports attaching volumes with type BlockDeviceVolume.
See
https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ec2.Instance.html#blockdevices
and
https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ec2.BlockDevice.html

### Use Case

CloudFormation supports attaching Volumes created separately from the Instance.  It would be nice to have the same support for the L2 CDK classes.

https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-instance.html#cfn-ec2-instance-volumes

### Proposed Solution
I believe this would involve adding a new property, `volumes` to [Instance](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-ec2.Instance.html#blockdevices) of a type that maps a device name to volume.
https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-mount-point.html

### Other

* [ ] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

**Comment by drmoore718** [comment]
I found a workaround (Python)

```
        volume = ec2.Volume(self, 'serverVolume',
            removal_policy=cdk.RemovalPolicy.RETAIN,
            ...
        )
        instance = ec2.Instance(self, 'serverInstance',
            instance_type=ec2.InstanceType('t3a.small'),
            ...
        )
        instance.node.default_child.volumes = [
            ec2.CfnInstance.VolumeProperty(
                device='/dev/sda1',
                volume_id=volume.volume_id,
            )
        ]
```

**Comment by njlynch** [comment]
Thanks for the feature request and work-around! I'm a bit surprised this is a gap in the Instance construct, but so it is.

I've tagged this as P1, which means it should be on our near-term roadmap.

We welcome community contributions! If you are able, we encourage you to [contribute](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md). If you decide to contribute, please start an engineering discussion in this issue to ensure there is a commonly understood design before submitting code. This will minimize the number of review cycles and get your code merged faster.

**Comment by vincent-turato** [comment]
I'd like to submit a pull request for this! I'll start an engineering discussion in this issue after I familiarize myself a bit more with the context here.

**Comment by luxaritas** [comment]
Still relevant

**Comment by rredpoppy** [comment]
wow, just wow!

**Comment by kumsmrit** [comment]
Hi @heuristics222

Thank you for raising this issue.

While this gets picked up, you can unblock yourself today without waiting for the L2 change — the L1 `CfnInstance` already supports this, so you can set it using the `@aws-cdk/cfn-property-mixins` [CfnInstancePropsMixin](https://docs.aws.amazon.com/cdk/api/v2/docs/@aws-cdk_cfn-property-mixins.aws_ec2.CfnInstancePropsMixin.html):

*Array merge note*: `volumes` is a list, so `ArrayMergeStrategy.append()` can be applied here that appends all volume elements. More details about different array merge strategies: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.ArrayMergeStrategy.html

```ts
import { aws_ec2 as ec2mixin } from '@aws-cdk/cfn-property-mixins';
import { ArrayMergeStrategy, PropertyMergeStrategy } from 'aws-cdk-lib';

.....
instance.with(new ec2mixin.CfnInstancePropsMixin({
  volumes: [{
    device: '/dev/sdf',
    volumeId: volume.volumeId,
  }],
}, {
  strategy: PropertyMergeStrategy.combine({ arrays: ArrayMergeStrategy.append() }),
}));
```

This should unblock you immediately; let us know if this works for you.
